src/odbc/src/statement.cpp

```diff
@@ -976,22 +1013,90 @@ SqlResult::Type Statement::InternalFreeResources(int16_t option) {
 }
 
 void Statement::Close() {
-  IGNITE_ODBC_API_CALL(InternalClose());
+  IGNITE_ODBC_API_CALL(InternalClose(false));
 }
 
-SqlResult::Type Statement::InternalClose() {
+SqlResult::Type Statement::InternalClose(bool ignoreErrors) {
   if (!currentQuery.get())
     return SqlResult::AI_SUCCESS;
 
   if (!currentQuery->DataAvailable()) {
     AddStatusRecord(SqlState::S24000_INVALID_CURSOR_STATE,
                     "No cursor was open");
-    return SqlResult::AI_ERROR;
+    return ignoreErrors ? SqlResult::AI_SUCCESS : SqlResult::AI_ERROR;
   }
 
   SqlResult::Type result = currentQuery->Close();
 
-  return result;
+  return ignoreErrors ? SqlResult::AI_SUCCESS : result;
+}
+
+void Statement::ExtendedFetch(SQLUSMALLINT orientation, SQLLEN offset, SQLULEN* rowCount, SQLUSMALLINT* rowStatusArray) {
+    IGNITE_ODBC_API_CALL(InternalExtendedFetch(orientation, offset, rowCount, rowStatusArray));
+}
+
+SqlResult::Type Statement::InternalExtendedFetch(SQLUSMALLINT orientation, SQLLEN offset,
+    SQLULEN* rowCount, SQLUSMALLINT* rowStatusArray) {
+  LOG_DEBUG_MSG("InternalExtendedFetch called with orientation " << orientation);
+  UNREFERENCED_PARAMETER(offset);
+
+  if (orientation != SQL_FETCH_NEXT) {
+    AddStatusRecord(SqlState::SHYC00_OPTIONAL_FEATURE_NOT_IMPLEMENTED,
+      "Only SQL_FETCH_NEXT FetchOrientation type is supported");
+    return SqlResult::AI_ERROR;
+  }
+
+  if (rowCount)
+    *rowCount = 0;
+
+  if (!currentQuery.get()) {
+    std::string errMsg = "Cursor is not in the open state.";
+    AddStatusRecord(SqlState::S24000_INVALID_CURSOR_STATE, errMsg);
+    return SqlResult::AI_ERROR;
+  }
+
+  // offset (aka FetchOffset) is ignored. Byte offsets are
+  // explicitly set to 0 to indicate that bind offsets are not supported
+  for (app::ColumnBindingMap::iterator it = columnBindings.begin();
+    it != columnBindings.end(); ++it) {
+    it->second.SetByteOffset(0);
+  }
+
+  SqlUlen fetched = 0;
+  SqlUlen errors = 0;
+
+  LOG_DEBUG_MSG("rowsetSize is " << rowsetSize);
+  for (SqlUlen i = 0; i < rowsetSize; ++i) {
+    for (app::ColumnBindingMap::iterator it = columnBindings.begin();
+      it != columnBindings.end(); ++it) {
+      it->second.SetElementOffset(i);
+    }
+
+    SqlResult::Type res = currentQuery->FetchNextRow(columnBindings);
+
+    if (res == SqlResult::AI_SUCCESS || res == SqlResult::AI_SUCCESS_WITH_INFO) {
+      ++fetched;
+    } else if (res != SqlResult::AI_NO_DATA) {
+      ++errors;
+    }
+
+    if (rowStatusArray) {
+      rowStatusArray[i] = SqlResultToRowResult(res);
+    }
+  }
+
+  if (rowCount) {
+    *rowCount = fetched;
+  }
+
+  if (fetched > 0) {
+    return errors == 0 ? SqlResult::AI_SUCCESS
+      : SqlResult::AI_SUCCESS_WITH_INFO;
+  }
+  
+  LOG_DEBUG_MSG("rowCount is " << rowCount << ", fetched is " << fetched
```

Comment:
Changed to check `rowCount` and report `*rowCount`. I have also moved this debug log so that it is reported whether errors occur or not, by moving it above the `fetched` check.